Orange/projection/pca.py

```diff
@@ -152,6 +152,35 @@ def partial_fit(self, data):
         return self
 
 
+class TruncatedSVD(SklProjector, _FeatureScorerMixin):
+    __wraps__ = skl_decomposition.TruncatedSVD
+    name = 'truncated svd'
+
+    def __init__(self, n_components=None, copy=True, whiten=False,
+                 svd_solver='auto', tol=0.0, iterated_power='auto',
+                 random_state=None, preprocessors=None, max_components=None):
+        super().__init__(preprocessors=preprocessors)
+        if n_components is not None and max_components is not None:
+            raise ValueError("n_components and max_components can not both be defined.")
+        # max_components limits the number of PCA components if the minimum
+        # shape of the X matrix (after preprocessing) is higher than
+        # max_components, so that sklearn does not always compute the full
+        # transform, which is faster and uses less memory for big data.
+        self.max_components = max_components
+        self.params = vars()
+
+    def fit(self, X, Y=None):
+        params = self.params.copy()
+        if params["n_components"] is None and self.max_components is not None:
+            # shape of X after preprocessing
+            # Add -1 to avoid error in scikit fit_transform
+            # ValueError: n_components must be < n_features (strict)
+            params["n_components"] = min(min(X.shape) - 1, self.max_components)
```

Comment:
Why, then, instead of `n_components` and `max_components` doing about the same thing not have just `n_components` and override it if shape too small?

E.g. what happens if `n_components=1e12` is passed and `X.shape == (3, 3)`?